### PR TITLE
koord-manager: refactor noderesource and add resource calculation extenders

### LINKED ISSUE #1060: [proposal] noderesource controller supports resource calculator extenders
**What is your proposal**:

The node-resource controller of the koord-manager should allow extensions for resource calculation.

**Why is this needed**:

The resource calculation algorithm in the node-resource controller should be extensible. For instance, a non-overcommit extender can be used for some nodes' batch resource calculation, while the other nodes use the default.

**Is there a suggested solution, if so, please add it**:

Add an extender framework in the node-resource controller.

### PR DESCRIPTION
**Pull request by saintube**
### Ⅰ. Describe what this PR does

Add the following extensions for the node-resource controller:

- NodeSync: allow the plugin to implement the node update check.
- ResourceReset: allow the plugin to reset the corresponding resources for degradation or colocation-disabled.
- ResourceCalculate: allow the plugin to implement different resource calculation algorithms other than the batch resource overcommitment.

### Ⅱ. Does this pull request fix one issue?

Fixes #1060.

### Ⅳ. Special notes for reviews

Just keep the GPU resources reconciliation since it will be refactored in the future.

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [X] All checks passed in `make test`

### CODE REVIEW

File: pkg/slo-controller/noderesource/resource_updater.go [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+	}
+	return false
+}
+
+func (r *NodeResourceReconciler) isGPULabelNeedSync(new, old map[string]string) bool {
+	return new[extension.LabelGPUModel] != old[extension.LabelGPUModel] ||
+		new[extension.LabelGPUDriverVersion] != old[extension.LabelGPUDriverVersion]
+}
+
+func (r *NodeResourceReconciler) updateGPUNodeResource(node *corev1.Node, device *schedulingv1alpha1.Device) error {
```
Comment (jasonliu747) [inline pkg/slo-controller/noderesource/resource_updater.go:113]: 计算GPU的部分，是不是可以抽离出来变成一个示范性的plugin，后续社区同学在开发的过程中也能过快速融入

Comment (jasonliu747) [reply pkg/slo-controller/noderesource/resource_updater.go:113]: 是的，我想的是既然都已经做成插件形式了，应抽尽抽😄

Comment (saintube) [reply pkg/slo-controller/noderesource/resource_updater.go:205]: 加上了

File: pkg/slo-controller/noderesource/batch_resource.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+func init() {
+	_ = RegisterNodeSyncExtender(BatchResourcePluginName, &BatchResourcePlugin{})
+	_ = RegisterNodePrepareExtender(BatchResourcePluginName, &BatchResourcePlugin{})
+	_ = RegisterResourceCalculateExtender(BatchResourcePluginName, &BatchResourcePlugin{})
+}
+
+const BatchResourcePluginName = "BatchResource"
+
+type BatchResourcePlugin struct{}
```
[the remark this answers is not in the document]
Comment (saintube) [reply pkg/slo-controller/noderesource/batch_resource.go]: done.

File: pkg/slo-controller/noderesource/device_resource_calculator.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
+/*
```
Comment (jasonliu747) [inline pkg/slo-controller/noderesource/device_resource_calculator.go:1]: 还是建议把 device_resource_calculator 也放在 plugins 里

File: pkg/slo-controller/noderesource/noderesource_controller.go [hunk 4]
```diff
@@ -34,21 +33,29 @@ import (
 	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
 	slov1alpha1 "github.com/koordinator-sh/koordinator/apis/slo/v1alpha1"
 	"github.com/koordinator-sh/koordinator/pkg/slo-controller/config"
+	"github.com/koordinator-sh/koordinator/pkg/slo-controller/noderesource/framework"
+	"github.com/koordinator-sh/koordinator/pkg/slo-controller/noderesource/plugins/batchresource"
 )
 
+func init() {
+	_ = framework.RegisterNodeSyncExtender(batchresource.PluginName, &batchresource.Plugin{})
```
Comment (zwzhang0107) [inline pkg/slo-controller/noderesource/noderesource_controller.go:41]: Register inside plugins/batchresource pkg

**Comment by saintube** [comment]
We keep the device resource calculation unchanged, which would be moved into a DP or add a standard implementation.

File: pkg/slo-controller/noderesource/noderesource_controller.go (hunk 4 above)
Comment (saintube) [reply pkg/slo-controller/noderesource/noderesource_controller.go:41]: done

**Pull request merged**